### PR TITLE
K8SPXC-856 feature to add volumes for sidecars

### PR DESCRIPTION
**Pull request by sridharlreddy**
[K8SPXC-856](https://jira.percona.com/browse/K8SPXC-856) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

This feature provides support for adding volumes to sidecar containers. Volumes can be added to sidecars using sidecarVolumes and sidecarPVCs on CR as below.

```    
    sidecars:
    - image: busybox
      command: ["/bin/sh"]
      args: ["-c", "while true; do trap 'exit 0' SIGINT SIGTERM SIGQUIT SIGKILL; done; "]
      name: my-sidecar-1
      volumeMounts:
      - mountPath: /volume1
        name: sidecar-volume-claim
      - mountPath: /secret
        name: sidecar-secret
      - mountPath: /config
        name: sidecar-config
    sidecarVolumes:
    - name: sidecar-secret
      secret:
        secretName: mysecret
    - name: sidecar-config
      configMap:
        name: myconfigmap
    sidecarPVCs:
    - apiVersion: v1
      kind: PersistentVolumeClaim
      metadata:
        name: sidecar-volume-claim
      spec:
        resources:
          requests:
            storage: 1Gi
        volumeMode: Filesystem
        accessModes:
          - ReadWriteOnce 
 
```

### CODE REVIEW

**Comment by spron-in** [comment]
Hello @sridharlreddy . We have a plan to add the same functionality in our Operator for MongoDB.
You can see the spec [here](https://docs.google.com/document/d/1d1N09FzfqQKiggG-4gzB3cvaLtEHd01jDmu0F15c3w0/edit).

What do you think about the approach in the spec? We want to have the similar approach in PXC.

**Comment by sridharlreddy** [comment]
> Hello @sridharlreddy . We have a plan to add the same functionality in our Operator for MongoDB.
> You can see the spec [here](https://docs.google.com/document/d/1d1N09FzfqQKiggG-4gzB3cvaLtEHd01jDmu0F15c3w0/edit).
> 
> What do you think about the approach in the spec? We want to have the similar approach in PXC.

Hi @spron-in ,

That's good to hear, I could not find any work being done on this, so I implemented the change for our needs and created the PR back for upstream.

For PXC, sidecars are already pass-through with ContainerSpec (env/volumeMounts are in ContainerSpec) and the limitation was outside ContainerSpec i.e at PodSpec and StatefulSetSpec. 
I extended the same to Volumes via VolumeSpec and PVCs via VolumeClaimTemplateSpec as passthrough to k8s.

I think we are trying to achieve the exact same requirement.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | passed |
| scheduled-backup | passed |
| haproxy | passed |
| recreate | passed |
| big-data | passed |
| upgrade-haproxy | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| self-healing | passed |
| limits | passed |
| monitoring-2-0 | passed |
| self-healing-chaos | passed |
| demand-backup | passed |
| smart-update | passed |
| affinity | passed |
| self-healing-advanced | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| self-healing-advanced-chaos | passed |
| pitr | passed |
| users | failed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| operator-self-healing-chaos | passed |
| demand-backup-encrypted-with-tls | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/975/commits/98fc9baab9eba2e307c7d5053f3b23ae47afcb4d
image: `perconalab/percona-xtradb-cluster-operator:PR-975-98fc9baa`

**Comment by spron-in** [comment]
@sridharlreddy thank you for the PR, it looks great.
But would be awesome to align it with what we wanted to do with MongoDB.

So the CR should look like this:
```
    sidecars:
    - image: busybox
      command: ["/bin/sh"]
      args: ["-c", "while true; do trap 'exit 0' SIGINT SIGTERM SIGQUIT SIGKILL; done; "]
      name: my-sidecar-1
      volumeMounts:
      - mountPath: /volume1
        name: sidecar-volume-claim
      - mountPath: /secret
        name: sidecar-secret
      - mountPath: /config
        name: sidecar-config
    volumes:
    - name: sidecar-secret
      secret:
        secretName: mysecret
    - name: sidecar-config
      configMap:
        name: myconfigmap
```

What I've done:
- renamed `sidecarVolumes` to just `volumes`
- removed `sidecarPVCs`. I really don't want to handle additional logic of PVCs in the Operator and just allow user to create the PVC separately. 


What do you think?

**Comment by sridharlreddy** [comment]
@egegunes Could you please help with the Test failure?

**Comment by sridharlreddy** [comment]
@spron-in there is **volumeSpec** attribute defined already, so it might confuse between this and additional volumes. 
For user created PVCs, same PVC will be shared and they have to be in RWX mode else just claim name cannot be used as same PVC is mounted to container across all the PODs/replicas of statefulset. This will be a different requirement than having a persistent volume for each pod.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| upgrade-haproxy | passed |
| big-data | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| scaling | passed |
| haproxy | passed |
| storage | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| security-context | passed |
| limits | passed |
| self-healing-chaos | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| self-healing-advanced-chaos | passed |
| affinity | passed |
| operator-self-healing | passed |
| one-pod | passed |
| auto-tuning | passed |
| operator-self-healing-chaos | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| scheduled-backup | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/975/commits/98fc9baab9eba2e307c7d5053f3b23ae47afcb4d
image: `perconalab/percona-xtradb-cluster-operator:PR-975-98fc9baa`

**Comment by spron-in** [comment]
> For user created PVCs, same PVC will be shared and they have to be in RWX mode as using claim name mounts same PVC to containers across all the PODs/replicas of statefulset. This will be a different requirement than having a persistent volume for each pod.

Yeah, agree here.

**Comment by egegunes** [comment]
LGTM. @spron-in what is your stand on field names?

**Comment by spron-in** [comment]
I would like to have smth like this:

```
    sidecars:
    - image: busybox
      command: ["/bin/sh"]
      args: ["-c", "while true; do trap 'exit 0' SIGINT SIGTERM SIGQUIT SIGKILL; done; "]
      name: my-sidecar-1
      volumeMounts:
      - mountPath: /secret
        name: sidecar-secret
      - mountPath: /config
        name: sidecar-config
      volumes:
      - name: sidecar-secret
        secret:
          secretName: mysecret
      - name: sidecar-config
        configMap:
          name: myconfigmap
      volumeSpec:
        persistentVolumeClaim:
          resources:
            requests:
              storage: 1Gi
          storageClassName: sc-name
          accessModes:
            - ReadWriteOnce 
```

1. First, I do not understand this:
```
      volumeMounts:
      - mountPath: /volume1
        name: sidecar-volume-claim
```
It is not possible to have multiple PVCs with the same name. So we should not be adding PVCs into volumeMounts and deal with this automatically. (see point 3)

2. I want to move volumes, volumeMounts and volumeSpec under `sidecars.[]` section, so that we have it distinguished.

3. Persistent Volume Claim should be created and mounted automatically to a sidecar.

4. ConfigMaps and Secrets are going to be handled through volumes and volumeMounts section.

**Comment by sridharlreddy** [comment]
Hi @spron-in 

1. This follows the statefulset spec, I think you are trying to compare with just bare podspec and might be confusing. 
    STS spec takes the volumeclaim template as input, creates a pvc for each pod from this template. pvc name defined in the template here is used as prefix followed by STS name and pod index. 
   Volumes for POD are auto-populated by STS and name of volume=claimName defined in template . So STS manages the this mapping and not to be specified under volumes.
  
2. Currently sidecar is directly mapped to ContainerSpec in k8s core. If you move volumes under sidecar, then you have to create a new type based on container spec and volume spec. Also volumes are defined at pod level and volumeMounts at container level, so it might create too many issues to combine them.

3. Statefulset manages PVCs based on volumeClaimTemplates defined under it

4. Yes.

Basically there are three kinds we are managing here
volumeMounts ==> At container level
volumes ==> At pod level. For STS, this does not include persistent volumes defined in volumeClaimTemplates. This includes only secrets, configmaps, emptDir, hostpath, ... 
volumeClaimTemplates ==> At statefulset level


I think it's better to discuss over a call, could you please let me know a suitable timeslot, will schedule one.

**Comment by sridharlreddy** [comment]
image

**Comment by sridharlreddy** [comment]
@spron-in What do you think?

**Comment by sridharlreddy** [comment]
@spron-in Should we skip PVCs or proceed with it?

**Comment by spron-in** [comment]
I'm okay to proceed with original PR with PVCs and your original CR layout.

**Comment by egegunes** [comment]
@qJkee could you take a look?

**Review by nonemax — APPROVED** [review]
lgtm

**Review by egegunes — APPROVED** [review]

**Comment by kalinkrustev** [comment]
Unfortunately this does not work, as reported here https://forums.percona.com/t/sidecarpvcs-metadata-pruning-issue-in-percona-mongodb-operator-deployment/23077

**Comment by gkech** [comment]
hello @kalinkrustev, the issue you are sharing from the percona forum is for percona mongodb operator, while here we have xtradb operator. On which one did you observe this behaviour?

**Comment by kalinkrustev** [comment]
I observed it for the xtradb, but the reason is the same as described in the link for the mongodb operator.

This seems to be the relevant place within the CRD: https://github.com/percona/percona-xtradb-cluster-operator/blob/de6913ed1659fe5b40d4ebec0812b6e5ede2332c/deploy/crd.yaml#L3169

**Comment by gkech** [comment]
> I observed it for the xtradb, but the reason is the same as described in the link for the mongodb operator.
> 
> This seems to be the relevant place within the CRD:
> 
> https://github.com/percona/percona-xtradb-cluster-operator/blob/de6913ed1659fe5b40d4ebec0812b6e5ede2332c/deploy/crd.yaml#L3169

Since this is a really old PR, I recommend to follow the process described here and open and issue:

https://github.com/percona/percona-xtradb-cluster-operator?tab=readme-ov-file#submitting-bug-reports
